### PR TITLE
Fix .to(cpu) for Storage

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138011

### CODE REVIEW

File: torch/_utils.py [hunk 1]
```diff
@@ -67,6 +67,17 @@ def _to(self, device, non_blocking=False):
     if self.device == device:
         return self
 
+    if device.type == "cpu":
+        pin_memory = non_blocking and self.device.type in (
+            "cuda",
+            torch._C._get_privateuse1_backend_name(),
+        )
```
Comment (mikaylagawarecki) [inline torch/_utils.py:74]: Following convention for tensors here https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/TensorConversions.cpp?fbclid=IwZXh0bgNhZW0CMTEAAR2nGqQndR7ZZk_X1azbVlPJHPXo9WbD1nA6e63VoOC6VGZSlsziYFJPjVk_aem_2ONazxgggrfNnWlfC6JIqA#L341-L342

File: test/test_torch.py [hunk 2]
```diff
@@ -5086,10 +5086,21 @@ def _get_tensors(**kwargs):
...
+    def test_storage_all_devices(self, devices, non_blocking):
         for device in devices:
-            t = torch.tensor((), device=device)
+            t = torch.randn(6, device=device)
             self.assertEqual(t.dtype, t.storage().dtype)
+            s = t.untyped_storage()
+            s_cpu = s.to(device='cpu', non_blocking=non_blocking)
+            if non_blocking:
+                event = torch.cuda.current_stream().record_event()
+                event.wait()
```
Comment (albanD) [inline test/test_torch.py:5098]: would a `torch.cuda.synchronize()` be simpler here?

File: torch/_utils.py [hunk 3]
```diff
@@ -67,6 +67,17 @@ def _to(self, device, non_blocking=False):
     if self.device == device:
         return self
 
+    if device.type == "cpu":
+        pin_memory = non_blocking and self.device.type in (
```
Comment (albanD) [inline torch/_utils.py:71]: You don't use this variable? :D

File: torch/storage.py [hunk 4]
```diff
@@ -273,9 +273,9 @@ def _to(self, dtype):
             storage = storage.clone()
         return storage
 
-    def to(
-        self, *, device: torch.device, non_blocking: _bool = False
-    ) -> Union[_StorageBase, TypedStorage]:
+    def to(self, *, device: Union[str, torch.device], non_blocking: _bool = False):
```
Comment (Skylion007) [inline torch/storage.py:276]:
```suggestion
    def to(self, *, device: torch._prims_common.DeviceLikeType, non_blocking: _bool = False):
```
Is what it should be

File: torch/storage.py [hunk 5]
```diff
@@ -273,9 +273,9 @@ def _to(self, dtype):
...
+        if isinstance(device, str):
+            device = torch.device(device)
```
Comment (Skylion007) [inline torch/storage.py:278]:
```suggestion
    def to(self, *, device: torch._prims_common.DeviceLikeType, non_blocking: _bool = False):
        if not isinstance(device, torch.device):
            device = torch.device(device)
```
Is what it should be

**Review by albanD — APPROVED** [review]
Sounds good!

**Pull request closed without merging**